### PR TITLE
Replace mcrypt by openssl

### PR DESCRIPTION
**Pull request by pmeulen**
Fix bug where the encryption options were not passed to the encryption class, added tests and cleanup the encryption interface.

Remove the mcrypt encryption class:
* The mcrypt extention has been removed in PHP 7.2
* The implementation uses a static IV, which is an unsafe cryptographic practice

Add openssl encryption class
- Allow algorithm to be selected to allow more modern encryption suited to be used, defaults to aes-128-cbc which should be available in any openssl currently in use is sufficient for the purpose of database encryption.
- Store the key id and cipher scheme that was used for encryption with the encrypted secret to allow key / cipher rollover
- Fix: Stop using a fixed IV to allow a counter mode cipher like GCM to be used. IV's are randomly generated per encrypted secret.

Add facilities to UserSecretStorage for migrating between encryption implementations.

Add a lot of tests

### CODE REVIEW

File: library/tiqr/Tiqr/UserSecretStorage/Encryption/Interface.php [hunk 1]
```diff
@@ -17,6 +17,8 @@
  * @copyright (C) 2010-2012 SURFnet BV
  */
 
+use Psr\Log\LoggerInterface;
```
Comment (MKodde) [inline library/tiqr/Tiqr/UserSecretStorage/Encryption/Interface.php]: Unused import? I could not find a reference in the web viewer.

File: library/tiqr/tests/Tiqr_UserSecretStorageTest.php [hunk 2]
```diff
@@ -5,16 +5,45 @@
 use PHPUnit\Framework\TestCase;
 use Psr\Log\LoggerInterface;
 
+class CustomEncryptionClass_2 implements Tiqr_UserSecretStorage_Encryption_Interface
```
Comment (MKodde) [inline library/tiqr/tests/Tiqr_UserSecretStorageTest.php]: Rule of thumb is to have 1 class per file. Given this is a test, it's excusable to add a test class to the existing test. But this gets messy real quickly in my opinion.

File: library/tiqr/tests/Tiqr_UserSecretStorageTest.php [hunk 3]
```diff
@@ -29,6 +58,159 @@ public function test_it_can_create_user_sercret_storage($type, $expectedInstance
...
+     * the same data
+     * @param string $type: file or pdo
+     * @param array $encryption_options: encryption/decryption options
+     * @param array $raw_user_secret: raw user secret data, array(userid=>secret), this is used to initialize the storage
+     *                                directly, without using the setSecret() method so it bypasses the encryption of the
+     *                                secret and the prefixing of the secret with the encryption type
+     *
+     * @return Tiqr_UserSecretStorage_Interface
+     */
+    public function setup_userSecretStorage(string $type, array $encryption_options = array(), array $raw_user_secret = array()) : Tiqr_UserSecretStorage_Interface
```
Comment (MKodde) [inline library/tiqr/tests/Tiqr_UserSecretStorageTest.php]: I hope this method is only used in this class. If so, please make it private.

File: library/tiqr/tests/Tiqr_UserSecretStorageTest.php [hunk 4]
```diff
@@ -29,6 +58,159 @@ public function test_it_can_create_user_sercret_storage($type, $expectedInstance
...
+            $type,
+            Mockery::mock(LoggerInterface::class)->shouldIgnoreMissing(),
+            $encryption_options
+        );
+    }
+
+    /**
+     * @dataProvider provideUserSecretStorageTypes
+     */
+    public function test_it_can_use_user_secret_storage_with_openssl_encryption(string $type)
```
Comment (MKodde) [inline library/tiqr/tests/Tiqr_UserSecretStorageTest.php:103]: This method might be a good candidate to deck out with a dataProvider that broadens the test coverage of this method to test different ciphers.

**Review by MKodde — COMMENTED** [review]
Nice work once again. I did take some getting used to that the PSR-12 standard was not kept. I try to use that whenever possible. But given the mix of Zend code style mixed with PSR-12, I can see where your naming decisions came from. 

My rule of thumb (or personal preference) is; when possible, all new code should stick to our chosen code style (PSR-12).

**Comment by pmeulen** [comment]
> Nice work once again. 

Thx @MKodde !

> I did take some getting used to that the PSR-12 standard was not kept. I try to use that whenever possible. But given the mix of Zend code style mixed with PSR-12, I can see where your naming decisions came from.

> My rule of thumb (or personal preference) is; when possible, all new code should stick to our chosen code style (PSR-12).

Your PSR-12 remark is especially about class names I think? Yeah, PSR-1 is from 2012, this library was started in 2010 and does not use namespaces. We could certainly change that. I don't see it as a high priority, but feel free to convince me otherwise :) I do value consistency within the project more than new code being PSR-12. Code should be clear and understandable. When we go to PSR-12, I want to change the whole library in one go.

I'll do some rework on the other points you mentioned.

File: library/tiqr/tests/Tiqr_UserSecretStorageTest.php (hunk 4 above)
Comment (pmeulen) [reply library/tiqr/tests/Tiqr_UserSecretStorageTest.php:103]: What I want to test here is that the openssl encryption class is instantiated correctly, and that the options array is passed to it correctly.

All the different ciphers are tested here:
https://github.com/Tiqr/tiqr-server-libphp/blob/cb45348483d2ae24ce3db1343f00ccf40acf92ab/library/tiqr/tests/Tiqr_UserSecretStorage_Encryption_OpenSSLTest.php#L47-L80

**Comment by MKodde** [comment]
I was mainly triggered by the method names actually. We usually only use the snake case in test cases (as that makes the longer method names in tests slightly more readable) method names are written in lowerCamelCase.

But I'm totally fine with updating that at a later stage in one go!

And yes, PSR-12 or the Doctrine coding standard is the way to go then.

`https://github.com/doctrine/coding-standard`

**Pull request merged**